biz.aQute.bndlib/src/aQute/bnd/osgi/PermissionGenerator.java

```diff
@@ -0,0 +1,147 @@
+				for (String namespace : builder.getProvideCapability().keySet()) {
+					if (!Processor.isDuplicate(namespace)) {
+						sb.append("(org.osgi.framework.CapabilityPermission \"").append(namespace)
+								.append("\" \"provide\")\n");
+					}
+				}
+				for (String namespace : builder.getRequireCapability().keySet()) {
+					if (!Processor.isDuplicate(namespace)) {
+						sb.append("(org.osgi.framework.CapabilityPermission \"").append(namespace)
+								.append("\" \"require\")\n");
+					}
+				}
+			}
+		},
+		PACKAGES {
+			@Override
+			public void generate(StringBuilder sb, Builder builder) {
+				if (builder.getImports() != null) {
+					for (PackageRef imp : builder.getImports().keySet()) {
+						if (!imp.isJava()) {
+							sb.append("(org.osgi.framework.PackagePermission \"");
+							sb.append(imp);
+							sb.append("\" \"import\")\n");
+						}
+					}
+				}
+				if (builder.getExports() != null) {
+					for (PackageRef exp : builder.getExports().keySet()) {
+						sb.append("(org.osgi.framework.PackagePermission \"");
+						sb.append(exp);
+						sb.append("\" \"export\")\n");
+					}
+				}
+			}
+		},
+		SERVICES {
+			@Override
+			public void generate(StringBuilder sb, Builder builder) {
+				for (String declaredService : getDeclaredServices(builder)) {
+					sb.append("(org.osgi.framework.ServicePermission \"").append(declaredService)
+							.append("\" \"register\")\n");
+				}
+				for (String referencedService : getReferencedServices(builder)) {
+					sb.append("(org.osgi.framework.ServicePermission \"").append(referencedService)
+							.append("\" \"get\")\n");
+				}
+			}
+		};
+
+		public abstract void generate(StringBuilder sb, Builder builder);
+	}
+
+	public static final String	KEY					= "permissions";
+
+	public static Set<String> getDeclaredServices(Builder builder) {
+		Set<String> declaredServices = new TreeSet<>();
+		for (Entry<String,Attrs> entry : builder.getProvideCapability().entrySet()) {
+			if (Processor.removeDuplicateMarker(entry.getKey()).equals("osgi.service")) {
+				Attrs attrs = entry.getValue();
+				String ifaces = attrs.get("objectClass");
+				if (ifaces != null) {
+					for (String iface : ifaces.split(",")) {
+						declaredServices.add(iface.trim());
+					}
+				}
+			}
+		}
+		return declaredServices;
+	}
+
+	private static final Pattern	FILTER_PATTERN	= Pattern
+															.compile("^\\(objectClass=([\\p{javaJavaIdentifierPart}\\.]+)\\)$");
+
+	public static Set<String> getReferencedServices(Builder builder) {
+		Set<String> referencedServices = new TreeSet<>();
+		for (Entry<String,Attrs> entry : builder.getRequireCapability().entrySet()) {
+			if (Processor.removeDuplicateMarker(entry.getKey()).equals("osgi.service")) {
+				Attrs attrs = entry.getValue();
+					String filter = attrs.get("filter:");
+					if (filter != null && !filter.isEmpty()) {
+					Matcher matcher = FILTER_PATTERN.matcher(filter);
+					if (matcher.matches()) {
```

Comment:
Since the filter expression can  be arbitrarily complex, we can only try to handle a reasonable subset. Right now you handle the single expression `(objectClass=com.foo.Service)`. We could also expand to handle trailing wildcards in the service name: `(objectClass=com.foo.*)`. A more complex thing to support would be the objectClass expression in an AND list: `(&(X)(objectClass=com.foo.*)(Y))` where X and Y are other expressions. This latter thing would be a fairly complex regex so we can leave out for now.